Added notes for 2024-04-15 (#8073)
dotnet/csharplang: meetings/2024/LDM-2024-04-15.md

# C# Language Design Meeting for April 15, 2024

## Agenda

> - [Non-enumerable collection types](Julien)

Reviewer: ```suggestion
- [Non-enumerable collection types]
```
We never put names in the notes

>   - (https://github.com/dotnet/csharplang/issues/7744) - championed issue

Reviewer: Fred usually links to the relevant section in the notes here, and then puts external links at the start of that section. Is the deviation from that deliberate?

  - (https://github.com/dotnet/csharplang/pull/7895) - requested details

> - Interceptors [location specifier](https://github.com/dotnet/csharplang/blob/interceptors-2024-01/proposals/interceptors-issues-2024-01.md#add-interceptslocationstring-locationspecifier) (Rikki)

Reviewer: ```suggestion
- Interceptors [location specifier](https://github.com/dotnet/csharplang/blob/interceptors-2024-01/proposals/interceptors-issues-2024-01.md#add-interceptslocationstring-locationspecifier)
```

> - [Relax `Add` requirementfo collection expression conversion](https://github.com/dotnet/csharplang/issues/8034)

Reviewer 2: ```suggestion
- [Relax `Add` requirement for collection expression conversions to types implementing `IEnumerable`](https://github.com/dotnet/csharplang/issues/8034)
```

## Non-enumerable collection types

> Should collections that  are not enumerable be usable in parameter collections, if they have a CollectionBuilder attribute that defines a single Create method?

Reviewer: ```suggestion
Should collections that  are not enumerable be usable in parameter collections, if they have a `CollectionBuilder` attribute that defines a single `Create` method?
```

> Should collections that  are not enumerable be usable in parameter collections, if they have a CollectionBuilder attribute that defines a single Create method?

Reviewer 2: ```suggestion
Should collections that are not enumerable be usable in collection expressions and `params` collections, if they have a `CollectionBuilder` attribute that defines a single `Create` method?
```

> In this case, the element type of the collection can be determined from the Create method if it is not already known.

Reviewer: ```suggestion
In this case, the element type of the collection can be determined from the `Create` method if it is not already known.
```

### Conclusion

These proposed changes were approved.

## Interceptors

### File location

Interceptors must identify the location in code where interception occurs. The issue is how to identify that location for use in the interceptor attribute.

Things that have been tried or considered for identifying the file:

* Absolute paths, per the preview, are bad for portability.
* Relative paths required the context be passed around or a project base path.

The proposed arguments to `[InterceptsLocation]` are a version number and an opaque data string. The opaque data string is not intended to be human readable. It is expected that this string will be inserted by a generator that retrieves it via a call into the interceptors API that requests an interceptible location.

The proposal is to create the opaque string by combining a checksum of the file contents, an integer location of the call in the syntax and a string file name for error reporting. Hand authoring interceptor locations is an unsupported scenario.

A version identifier is included to allow future change to be made in the calculation of the data string.

### Other public API changes

> The team also mentioned an that there is an experimental public API that lets you determine whether syntax is being intercepted and whose intercepting. This might be used in an analyzer, for example.

Reviewer: ```suggestion
The team also mentioned that there is an experimental public API that lets you determine whether syntax is being intercepted and who is intercepting. This might be used in an analyzer, for example.
```

### Conclusion

The proposed approach to identifying the file location was approved, including the proposed handling of version.

> ## Relax `Add` requirement for collection expression conversion

Reviewer 2: ```suggestion
## Relax `Add` requirement for collection expression conversions to types implementing `IEnumerable`
```

https://github.com/dotnet/csharplang/issues/8034

We introduced a breaking change in 17.10 which had a larger impact than anticipated. The proposal is to resolve that breaking change. It occurs when:

> * Types that implement IEnumerable but not IEnumerable<T> and have a strongly-typed Add(T).

Reviewer: ```suggestion
* Types that implement `IEnumerable` but not `IEnumerable<T>` and have a strongly-typed `Add(T)`.
```

> * Types that implement IEnumerable<T> and have a strongly-typed Add(U) where U is implicitly convertible to T. This is a generic form of category 1.

Reviewer: ```suggestion
* Types that implement `IEnumerable<T>` and have a strongly-typed `Add(U)` where `U` is implicitly convertible to `T`. This is a generic form of category 1.
```

> * Types that implement IEnumerable<T> and have a strongly-typed Add(U) where U and T are unrelated.

Reviewer: ```suggestion
* Types that implement `IEnumerable<T>` and have a strongly-typed `Add(U)` where `U` and `T` are unrelated.
```

The breaking change added in 17.10 was the additional restriction:

> * Add needs to have one parameter of the iterator type - breaking change.

Reviewer: ```suggestion
* `Add` needs to have one parameter of the iterator type
```

> The proposal is to keep the Add method, still require a single Add parameter but no statement about the type of the argument.

Reviewer: ```suggestion
The proposal is to keep the `Add` method, still require a single `Add` parameter but no statement about the type of the argument.
```
